cdap-watchdog/src/main/java/co/cask/cdap/logging/framework/AvroFileManager.java

```diff
@@ -0,0 +1,248 @@
+import org.apache.avro.Schema;
+import org.apache.avro.file.DataFileWriter;
+import org.apache.avro.generic.GenericDatumWriter;
+import org.apache.avro.generic.GenericRecord;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.twill.filesystem.Location;
+import org.apache.twill.filesystem.LocationFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Class including logic for getting avro files to write to. Used by {@link CDAPLogAppender}
+ */
+class AvroFileManager {
+  private static final Logger LOG = LoggerFactory.getLogger(AvroFileManager.class);
+
+  private final long maxLifetimeMillis;
+  private final Map<LogPathIdentifier, LogFileOutputStream> fileMap;
+  private final Location logsDirectoryLocation;
+  private final FileMetaDataManager fileMetaDataManager;
+  private final int syncIntervalBytes;
+  private final Schema schema;
+
+  AvroFileManager(long maxFileLifetimeMs, int syncIntervalBytes, Schema schema,
+                  FileMetaDataManager fileMetaDataManager,
+                  LocationFactory locationFactory) {
+    this.maxLifetimeMillis = maxFileLifetimeMs;
+    this.syncIntervalBytes = syncIntervalBytes;
+    this.schema = schema;
+    this.fileMetaDataManager = fileMetaDataManager;
+    this.logsDirectoryLocation = locationFactory.create("logs");
+    this.fileMap = new HashMap<>();
+  }
+
+  /**
+   * Get an avro file for the give logging context and timestamp - return an already open file,
+   * or create and return a new file or rotate a file based on time and return.
+   * @param logPathIdentifier
+   * @param timestamp
+   * @return LogFileOutputStream
+   * @throws IOException
+   */
+  public LogFileOutputStream getAvroFile(LogPathIdentifier logPathIdentifier, long timestamp) throws IOException {
+    LogFileOutputStream logFileOutputStream = fileMap.get(logPathIdentifier);
+    // If the file is not open then set reference to null so that a new one gets created
+    if (logFileOutputStream == null) {
+      logFileOutputStream = createAvroFile(logPathIdentifier, timestamp);
+    }
+    // rotate the file if needed (time has passed)
+    logFileOutputStream = rotateFile(logFileOutputStream, logPathIdentifier, timestamp);
+    return logFileOutputStream;
+  }
+
+  private LogFileOutputStream createAvroFile(LogPathIdentifier identifier, long timestamp) throws IOException {
+    Location location = getLocation(identifier, timestamp);
+    LOG.info("Creating Avro file {}", location);
+    LogFileOutputStream logFileOutputStream = new LogFileOutputStream(location);
+    try {
+      fileMetaDataManager.writeMetaData(identifier, timestamp, location);
+    } catch (Throwable e) {
+      closeAndDelete(logFileOutputStream);
+      throw new IOException(e);
+    }
+
+    fileMap.put(identifier, logFileOutputStream);
```

Comment:
so we call `clear` when `close` of the LogFileManager is called, but that close is only called on CDAPLogAppender stop.  we update the map on rotate, but when there is a log context where there aren't any new messages, in our current logic we don't close it, its in the map. i am thinking when we perform flush on all files, we can check if the file's start time is older than TTL and close those files and remove them from the map. does that sound good ?